src/main/java/org/myrobotlab/service/NeoPixel.java

```diff
@@ -26,501 +26,829 @@
+
+  /**
+   * RGB or RGBW supported 3 RGB 4 RGBW
+   */
+  protected int pixelDepth = 3;
+  /**
+   * current selected red value
+   */
+  
+  protected int red = 0;
+
+  /**
+   * currently selected animation
+   */
+  protected String currentAnimation;
+
+  /**
+   * speed of an animation in fps
+   */
+  protected int speedFps = 10;
+  
+  /**
+   * map between speed and wait time in ms for those animations that
+   * need it (onboard)
+   */
+  protected MapperLinear fpsToWaitMs = new MapperLinear(1, 30, 1000, 30);
 
   public NeoPixel(String n, String id) {
     super(n, id);
-    animationSettings.put(NEOPIXEL_ANIMATION_STOP, new boolean[] { false, false });
-    animationSettings.put(NEOPIXEL_ANIMATION_COLOR_WIPE, new boolean[] { true, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_LARSON_SCANNER, new boolean[] { true, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_THEATER_CHASE, new boolean[] { true, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_THEATER_CHASE_RAINBOW, new boolean[] { false, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_RAINBOW, new boolean[] { false, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_RAINBOW_CYCLE, new boolean[] { false, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_FLASH_RANDOM, new boolean[] { true, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_IRONMAN, new boolean[] { true, true });
-    subscribeToRuntime("registered");
-  }
-
-  public void onRegistered(Registration s) {
-    refreshControllers();
-    broadcastState();
+    animationRunner = new AnimationRunner();
+    Double d = fpsToWaitMs.calcOutput(speedFps);
+    animationWaitMs = d.intValue();
   }
 
-  public List<String> refreshControllers() {
-    controllers = Runtime.getServiceNamesFromInterface(NeoPixelController.class);
-    return controllers;
-  }
+  @Override
+  public void attach(Attachable service) throws Exception {
 
-  // @Override
-  public NeoPixelController getController() {
-    return controller;
-  }
+    if (SpeechSynthesis.class.isAssignableFrom(service.getClass())) {
+      attachSpeechSynthesis((SpeechSynthesis) service);
+      return;
+    }
 
-  public String getControllerName() {
-    String controlerName = null;
-    if (controller != null) {
-      controlerName = controller.getName();
+    if (NeoPixelController.class.isAssignableFrom(service.getClass())) {
+      attachNeoPixelController((NeoPixelController) service);
+      return;
     }
-    return controlerName;
+    warn(String.format("%s.attach does not know how to attach to a %s", this.getClass().getSimpleName(), service.getClass().getSimpleName()));
   }
 
-  public boolean isAttached() {
+  @Override
+  public void attachNeoPixelController(NeoPixelController neoCntrlr) {
+
     if (controller != null) {
-      if (((Arduino) controller).getDeviceId((Attachable) this) != null) {
-        isAttached = true;
-        return true;
+      if (controller.equals(neoCntrlr.getName())) {

```

Comment:
this will blow a null pointer exception if neoCntrlr is null.

Reply:
fixed